pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KindCalicoNodeStatus     = "CalicoNodeStatus"
+	KindCalicoNodeStatusList = "CalicoNodeStatusList"
+)
+
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CalicoNodeStatusList is a list of CalicoNodeStatus resources.
+type CalicoNodeStatusList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []CalicoNodeStatus `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type CalicoNodeStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   CalicoNodeStatusSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status CalicoNodeStatusStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// CalicoNodeStatusSpec contains the specification for a CalicoNodeStatus resource.
+type CalicoNodeStatusSpec struct {
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
+	Classes []NodeStatusClassType `json:"classes,omitempty"`
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
```

Comment:
Wondering if we should have calico/node create one of these as part of its startup logic, and then users can simply `kubectl edit` it in order to enable / configure it.

If we default to `0s` here, then we won't start reporting until asked for. WDYT?

Reply:
I'm not very keen on that. Because
- a large cluster may create loads of NodeStatus objects but actually users are not going to use them at all. This resource is still more or less for debugging purpose. 
- User may just want to focus on one node but get lost on hundreds of objects listed with a `list` api call. 
- It is easier to create a resource correctly than to patch/edit it. Also easier to document as well.

Reply:
Agree more or less with the first two bullet points, but strongly disagree with ease of documentation. Easier to follow documentation was my main reason for suggesting this.

All of our documentation is going to have to be written in the form "If this object exists, then modify it. If it doesn't exist, create it".

Which is somewhat confusing

Reply:
This needs to be a `*int` if we want to perform defaulting on it - otherwise the code will see it as `0` and assume the user disabled refresh.

Reply:
I think the documentation point depends on what we consider the mainline use case to be.  If it is that these resources are only used when there is a specific need to investigate or understand something, then the documentation could say "Create this object, and read it back to see node debugging information.  When you've resolved the issue concerned, delete the object again."

Given that this object's identity is NOT the node name, and that there is only one Interval field per object, I feel that's the right model.  For example, suppose there is an existing CalicoNodeStatus for node MYNODE and classes Grok and Frobnicate: you can still create a new CalicoNodeStatus (called e.g. "mynode-bgp36") for node MYNODE to monitor BIRD stuff.

Reply:
BTW - and I don't know if this is even possible - but did you consider a model where status is reported as a status subresource of the Calico Node object?